stdlib/asm/crypto/stark/public_inputs.masm

```diff
@@ -30,16 +30,59 @@ export.load
         adv_loadw
         hperm
     end
+     
+    # 2) Compute the number of digests we have to load. The digests are the program hash
+    #    and kernel procedures digests.
+    exec.constants::get_num_kernel_procedures
+    add.1
 
-    # 2) Load the program digest
-    adv_loadw
+    # 3) Load the program hash and kernel procedures digests.
+    #    We need one call to the RPO permutation per 2 digests, thus we compute the division
+    #    with remainder of the number of digests by 2. If the remainder is 1 then we need
+    #    to pad with the zero word, while we don't need to pad otherwise.
+    u32divmod.2
+    push.0 eq
+    if.true
+        dup
+        movdn.14
+        push.0
+        neq
+        while.true
+            adv_pipe
+            hperm
+            movup.13
+            sub.1
+            movdn.13
+            dup.13
+            push.0
+            neq
+        end
```

Comment:
Can you sprinkle some comments that track the state of the stack in here (and in the very similar `else` branch)?

Reply:
Yep, added